src/lib.rs

```diff
@@ -136,6 +136,33 @@ impl SmolStr {
             _ => false,
         }
     }
+
+    fn from_char_iter<I: iter::Iterator<Item = char>>(mut iter: I) -> SmolStr {
+        let (min_size, _) = iter.size_hint();
+        if min_size > INLINE_CAP {
```

Comment:
That's not entirely correct (as inline CAP is in bytes), but I think good enough, as we don't guarantee that the repr is canonical yet (there might be equivalent inline and non-inline string). 

However, at some point we should make such guarantee (to speedup comparisons)

Reply:
If the iterator lies with its size_hint, and says the minimum number of items is greater than the actual number of items yielded, we can end up with a short string on the heap.

But if it isn't lying, this check is correct. INLINE_CAP is in bytes, and min_size is in chars, but every char takes at least one byte, therefore if there are min_size chars, and min_size > INLINE_CAP, those chars will encode to more than INLINE_CAP bytes.